pkg/common/util/dial.go

```diff
@@ -0,0 +1,79 @@
+package util
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// BlockingDial is a helper method to dial the given address, using optional TLS credentials,
+// and blocking until the returned connection is ready. If the given credentials are nil, the
+// connection will be insecure (plain-text).
+// (This function was borrowed from https://github.com/fullstorydev/grpcurl/blob/master/grpcurl.go)
+func BlockingDial(ctx context.Context, network, address string, creds credentials.TransportCredentials, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
+	// grpc.Dial doesn't provide any information on permanent connection errors (like
+	// TLS handshake failures). So in order to provide good error messages, we need a
+	// custom dialer that can provide that info. That means we manage the TLS handshake.
+	result := make(chan interface{}, 1)
+
+	writeResult := func(res interface{}) {
+		// non-blocking write: we only need the first result
+		select {
+		case result <- res:
+		default:
+		}
+	}
+
+	dialer := func(address string, timeout time.Duration) (net.Conn, error) {
+		ctx, cancel := context.WithTimeout(ctx, timeout)
+		defer cancel()
+
+		conn, err := (&net.Dialer{Cancel: ctx.Done()}).Dial(network, address)
+		if err != nil {
+			writeResult(err)
+			return nil, err
+		}
+		if creds != nil {
+			conn, _, err = creds.ClientHandshake(ctx, address, conn)
+			if err != nil {
+				writeResult(err)
+				return nil, err
+			}
+		}
+		return conn, nil
+	}
+
+	// Even with grpc.FailOnNonTempDialError, this call will usually timeout in
+	// the face of TLS handshake errors. So we can't rely on grpc.WithBlock() to
+	// know when we're done. So we run it in a goroutine and then use result
+	// channel to either get the channel or fail-fast.
+	go func() {
+		opts = append(opts,
+			grpc.WithBlock(),
```

Comment:
It might be desirable to keep the dial non-blocking here...? I am wondering if, instead of switching to a blocking dial, we can take a logger and print out errors that the dialer encounters with either Dial or ClientHandshake results... Maybe this whole thing can change to something a little more heavyweight, with a dedicated struct and interface? Something like:
```go
config := util.GRPCDialConfig{
     Log: logger_here,
     Creds: tls_credentials_here,
     Opts: []grpc.DialOption,
}

dialer := util.NewGRPCDialer(config)
conn, err := dialer.Dial(ctx, addr)
```

If we did that, perhaps this would also warrant a dedicated package.

Reply:
Makes sense to me. I will do that.